packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -816,6 +816,81 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
         }
     }
 
+    /**
+     * @inheritdoc IAssetManager
+     */
+    function claimKghReward(address validator, uint256[] calldata tokenIds) external {
+        if (validator == address(0)) revert ZeroAddress();
+        if (tokenIds.length == 0) revert NotAllowedZeroInput();
+
+        Vault storage vault = _vaults[validator];
+        KghDelegator storage kghDelegator = vault.kghDelegators[msg.sender];
+        if (kghDelegator.kghNum < tokenIds.length) revert InvalidTokenIdsInput();
+
+        // Auto compounding is not applied for claim function.
+        uint128 claimedRewards = _claimBoostedReward(validator, msg.sender, tokenIds.length, false);
+
+        // Burn kroShares to claim the base reward.
+        uint128 kroSharesToBurn;
+        for (uint256 i = 0; i < tokenIds.length; ) {
+            uint128 kroShares = kghDelegator.shares[tokenIds[i]].kro -
+                _convertToKroShares(validator, KGH_MANAGER.totalKroInKgh(tokenIds[i]));
+            if (kroShares == 0) revert ShareNotExists();
+
+            unchecked {
+                kroSharesToBurn += kroShares;
+
+                ++i;
+            }
+        }
+
+        uint128 kghBaseReward = _convertToKroAssets(validator, kroSharesToBurn);
+
+        unchecked {
+            claimedRewards += kghBaseReward;
+            vault.asset.totalKroShares -= kroSharesToBurn;
+            vault.asset.totalKro -= claimedRewards;
+        }
+
+        ASSET_TOKEN.safeTransfer(msg.sender, claimedRewards);
+
+        emit KghRewardClaimed(validator, msg.sender, claimedRewards, kroSharesToBurn);
+    }
+
+    /**
+     * @notice Claim the boosted reward of the delegator.
+     *
+     * @param validator     Address of the validator.
+     * @param delegator     Address of the delegator.
+     * @param kghNum        The number of KGH to claim the boosted reward.
+     * @param isCompounding Flag to auto compound the boosted reward.
+     *
+     * @return The amount of the claimed boosted reward.
+     */
+    function _claimBoostedReward(
+        address validator,
+        address delegator,
+        uint256 kghNum,
```

Comment:
I missed this part to use it, the below code should be `uint128 totalBoostedReward = kghNum * ...` not `uint128 totalBoostedReward = kghDelegator.kghNum *`, so it's needed.

Reply:
We need to claim all the accumulated boosted reward here always, because `rewardPerKghPaid` is updated for all the tokens, not for each token.
I think it should be removed and we need to use `kghDelegator.kghNum`

Reply:
Oh then we cannot do this for part of KGHs we have. I'll fix it, thanks!